src/join.rs

```diff
@@ -155,6 +155,61 @@ pub trait Join {
         JoinIter::new(self)
     }
 
+    /// Optionally gets a `Join` if it exists.
+    ///
+    /// ```
+    /// # use specs::prelude::*;
+    /// # #[derive(Debug, PartialEq)]
+    /// # struct Pos { x: i32, y: i32 } impl Component for Pos { type Storage = VecStorage<Self>; }
+    /// # #[derive(Debug, PartialEq)]
+    /// # struct Vel { x: i32, y: i32 } impl Component for Vel { type Storage = VecStorage<Self>; }
+    /// struct ExampleSystem;
+    /// impl<'a> System<'a> for ExampleSystem {
+    ///     type SystemData = (
+    ///         WriteStorage<'a, Pos>,
+    ///         ReadStorage<'a, Vel>,
+    ///     );
+    ///     fn run(&mut self, (mut positions, velocities): Self::SystemData) {
+    ///         for (mut position, maybe_velocity) in (&mut positions, velocities.maybe()).join() {
+    ///             if let Some(velocity) = maybe_velocity {
+    ///                 position.x += velocity.x;
+    ///                 position.y += velocity.y;
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let mut world = World::new();
+    ///     let mut dispatcher = DispatcherBuilder::new()
+    ///         .with(ExampleSystem, "example_system", &[])
+    ///         .build();
+    ///
+    ///     dispatcher.setup(&mut world.res);
+    ///
+    ///     let e1 = world.create_entity()
+    ///         .with(Pos { x: 0, y: 0 })
+    ///         .with(Vel { x: 5, y: 2 })
+    ///         .build();
+    ///
+    ///     let e2 = world.create_entity()
+    ///         .with(Pos { x: 0, y: 0 })
+    ///         .build();
+    ///
+    ///     dispatcher.dispatch(&mut world.res);
+    ///
+    ///     let positions = world.read_storage::<Pos>();
+    ///     assert_eq!(positions.get(e1), Some(&Pos { x: 5, y: 2 }));
+    ///     assert_eq!(positions.get(e2), Some(&Pos { x: 0, y: 0 }));
+    /// }
+    /// ```
+    fn maybe(self) -> MaybeJoin<Self>
```

Comment:
Iterating only over a `MaybeJoin` would iterate over 4,294,96,7296 `None` values. Don't know how to fix it, but this PR doesn't qualify for a merge before it's solved.

Reply:
I'm not sure it needs to be solved to be honest.  This would only happen if every element in a join tuple was a MaybeJoin, in which case we should probably just tell people "don't do that."

Reply:
As a slightly more user friendly alternative, we could store a variable in the `MaybeJoin` iterator called "count" and if count goes to max print a warning: "MaybeJoin iterated over all indices.  Do you have a join iterator consisting exclusively of MaybeJoin?  Consider refactoring this code to avoid performance problems."

Reply:
The only other alternative to this I can think of is to require `MaybeJoin` to store a `EntitiesRes` reference somehow so that it will only iterate over valid indices.

Reply:
@torkleyy Do you find any of my ideas here satisfactory?

Reply:
I'm not quite sure. Let's use a warning in the docs for now.

Reply:
A quicker way (and more completely covering way) is to just add 
```
if idx as usize == MAX_EID - 1 {
    println!("WARNING: Iterated through all indices ({} indices), you might've made a join with all `MaybeJoin`s, which is unbounded in length", MAX_EID);
}
```
But that would have a more significant cost.